cdist/conf/type/__package_apt/gencode-remote

```diff
@@ -66,25 +60,29 @@ aptget="DEBIAN_FRONTEND=noninteractive apt-get --quiet --yes -o Dpkg::Options::=
 case ${state_should}
 in
     present)
-        if test "${state_is}" = 'present'
+        if test -n "${version_should}" -a "${version_should}" != "${version_is}"
+        then
+            # other version should be installed
+            name="${name}=${version_should}"
+            aptget="${aptget:?} -o APT::Get::allow-downgrades=true"
+        elif test "${state_is}" = 'present'
         then
-            if test -z "${version}" || test "${version}" = "${version_is}"
+            if test "${auto_state}" = 'auto'
             then
-                if test "${auto_state}" = 'auto'
-                then
-                    # package is installed, but marked auto -> mark it manual so
-                    # that it doesn't get removed by apt-get autoremove.
-
-                    echo 'head -n1 "$(command -v apt-mark)" | grep -qvF python \'
-                    # new implementation (Debian >= 7)
-                    echo "&& apt-mark manual '${name}' \\"
-                    # old Python implementation (Debian 5-7)
-                    echo "|| apt-mark unmarkauto '${name}'"
-                    # NOTE: older versions don't need to be supported because
-                    # they will always be manual
-                fi
-                exit 0
+                # package is installed, but marked auto -> mark it manual so
+                # that it doesn't get removed by apt-get autoremove.
+
+                echo 'head -n1 "$(command -v apt-mark)" | grep -qvF python \'
+                # new implementation (Debian >= 7)
+                echo "&& apt-mark manual '${name_is}' \\"
+                # old Python implementation (Debian 5-7)
+                echo "|| apt-mark unmarkauto '${name_is}'"
```

Comment:
This part is unchanged from #781.

`grep -v` just inverts the exit code (`0` if no match was found).  
I felt like it makes more sense to have the "new" way as the success branch and fall back to the "old" way (which is still supported but prints a warning) if either the rudimentary check or the "new way" fails.

This block could be converted into an `if/else`, but it needs to be executed on the target. What would be the advantage?

Reply:
> This part is unchanged from #781.

Yeah, saw it, but I thought that... okay, fine, but now... :roll_eyes: 

> grep -v just inverts the exit code

That's bit confusing, because it's not that common practice. YMMV.

> This block could be converted into an if/else, but it needs to be executed on the target. What would be the advantage?

Mostly because of [SC2015](https://github.com/koalaman/shellcheck/wiki/SC2015).

IMHO, `foo && bar || baz` is just bad practice, no matter where it's executed or is it generated code or straight exec.

Reply:
> > grep -v just inverts the exit code
> That's bit confusing, because it's not that common practice. YMMV.

Hmm, okay. I could replace it with `! grep -qF` if you find that more legible?

> > This block could be converted into an if/else, but it needs to be executed on the target. What would be the advantage?
> > Mostly because of SC2015.
> IMHO, foo && bar || baz is just bad practice, no matter where it's executed or is it generated code or straight exec.

Ah, SC2015. I am aware of this detail but I took it as a "nice extra" if the new way fails for some reason, it automatically falls back to the old way.

Reply:
> Hmm, okay. I could replace it with ! grep -qF if you find that more legible?

Totally.

> I am aware of this detail but I took it as a "nice extra" if the new way fails for some reason, it automatically falls back to the old way.

Ahaa, clever hack.

So, all this is one uncommon approach with clever hack. I'm totally cool with both if you add comment about your motivation. Then we have some kind of safety net for the future if someone next is looking at the code and starts to scratch their head :wink:

Reply:
@4nd3r While writing the documentation I noticed that the clever hack is probably unnecessary.
The [implementation](https://salsa.debian.org/apt-team/apt/-/blob/2.3.11/cmdline/apt-mark.cc#L46-140) of the marking code is straight forward and, apart from writing to stdout, behaves exactly the same.
So I dropped the ternary anyway and replaced it with an `if/else`.